beginner_source/profiler_tutorial.py

```diff
@@ -0,0 +1,205 @@
+model(input, mask) #warm up
+
+with profiler.profile(with_stack=True, profile_memory=True) as prof:
+    out, idx = model(input, mask)
+
+print(prof.key_averages(group_by_stack_n=5).table())
+
+
+######################################################################
+# Improve time performance
+# ^^^^^^^^^^^^^^^^^^^^^^^^^
+# That worked, the memory footprint has halved - but while the time
+# consumed has reduced a bit (previously 283ms), it’s still too high at
+# 194ms. Turns out copying a matrix from CUDA to CPU is pretty expensive!
+# The ``aten::copy_`` operator in ``forward (11)`` copies ``mask`` to CPU
+# so that it can use the NumPy ``argwhere`` function. We could eliminate
+# the need to copy into a NumPy array if we use a ``torch`` function
+# ``nonzero()`` here instead.
+#
+
+class CustomLinearLite(torch.nn.Linear):
+    def __init__(self, in_features: int, out_features: int, bias: bool = True) -> None:
+        super().__init__(in_features, out_features, bias)
+
+    def forward(self, input, mask):
+        # with profiler.record_function("LINEAR PASS"):
+            out = super().forward(input)
+
+        # with profiler.record_function("INDEX SCORE"):
+            threshold = out.sum(axis=1).mean()
+            hi_idx = (mask > threshold).nonzero(as_tuple=True)
+
+            return out, hi_idx
+
+
+model = CustomLinearLite(500, 10).cuda()
+input = torch.rand(128, 500).cuda()
+mask = torch.rand((500, 500, 500), dtype=torch.float).cuda()
+
+model(input, mask) #warm up
+
+with profiler.profile(with_stack=True, profile_memory=True) as prof:
+    out, idx = model(input, mask)
+
+print(prof.key_averages(group_by_stack_n=5).table())
+
+
+######################################################################
+# Replacing ``np.argwhere`` with ``torch.nonzero`` reduced the total
+# runtime to about 36ms (from 195ms) by eliminating the need to copy to
+# CPU and instead running on the GPU.
+#
+
+
+######################################################################
+# Caveat
+# ^^^^^^^^^^^^^^
+#
+# You might notice introducing ``torch.nonzero`` in our module has shown a
+# new memory footprint of 2.59 Gb that wasn’t in the stacktrace when we
+# used ``np.argwhere``. This is because Profiler tracks only operations in
+# ``torch``; it is likely that ``np.argwhere`` occupied a similar amount
+# of memory on the CPU but that will not be reported in the trace here.
+#
+
+
+######################################################################
+# Some properties of Profiler:
+# ~~~~~~~~~~~~~~~~~
+#
+# Thread local
+# ^^^^^^^^^^^^^^
+#
+# Profiler runs in the same thread as the operation but it will also
+# profile child operators that might run in another thread.
+# Concurrently-running profilers will be scoped to their own thread to
+# prevent mixing of results.
+#
+# Module-level, not low-level system profiling
+# ^^^^^^^^^^^^^^
+#
+# PyTorch profiler only reports runtime of PyTorch functions, and not at
+# the lower-level system operations.
+#
```

Comment:
we need some conclusion, e.g. "we considered an example and showed how to use profiler to improve time and memory consumption of the model. For more examples, please check profiler recipe/profiler [links]"